rtloader/common/diagnoseutils.c

```diff
@@ -0,0 +1,98 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2019-present Datadog, Inc.
+#include "diagnoseutils.h"
+#include "stringutils.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+size_t get_diagnoses_mem_size(Py_ssize_t numDiagnoses, PyObject *diagnoses_list)
+{
+    Py_ssize_t idx = 0;
+
+    // Calculate and allocate buffer size
+    size_t bufferSize = sizeof(diagnosis_set_t) + (numDiagnoses * sizeof(diagnosis_t));
+    for (idx = 0; idx < numDiagnoses; idx++) {
+        PyObject *diagnosisObj = PyList_GetItem(diagnoses_list, idx); // borrowed ref
+        if (diagnosisObj == NULL) {
+            return 0;
+        }
+
+        bufferSize += attr_as_string_size(diagnosisObj, "name");
+        bufferSize += attr_as_string_size(diagnosisObj, "diagnosis");
+        bufferSize += attr_as_string_size(diagnosisObj, "description");
+        bufferSize += attr_as_string_size(diagnosisObj, "remediation");
+        bufferSize += attr_as_string_size(diagnosisObj, "raw_error");
+    }
+
+    return bufferSize;
+}
+
+int serialize_diagnoses(Py_ssize_t numDiagnoses, PyObject *diagnoses_list, diagnosis_set_t *diagnoses,
```

Comment:
This is very cool, but if we're serializing things anyway in a byte buffer anyway, can't we just pass JSON?

Reply:
I'm not proposing to change anything in CLI, only the Python/Go interface: serialize Python objects to JSON on the Python side (or in rtloader if you must, see `as_yaml` for an example), and parse into proper Go type in python/check.go `GetDignoses`. rtloader does not need to know what Diagnosis is beyond that it's a byte blob.

Reply:
Fascinating suggestion which I have not recognized in the beginning. Yes, indeed it would remove tons of complexity I have introduced around various string serialization and also make the rtloader nearly completely decoupled from either side, as you have fairly noted. I have attempted to do it. I spent half a day but was not able to figure out how to make sure that the Python Diagnosis class/object is serializable as YAML (took me a while to even find the root cause - luckily I have found PyErr_Print function).

The problem is that the YAML dumper we use cannot dump custom classes. I fail with the following stack
```
Traceback (most recent call last):
  File "C:\Program Files\Datadog\Datadog Agent\embedded3\lib\site-packages\yaml\__init__.py", line 253, in dump
    return dump_all([data], stream, Dumper=Dumper, **kwds)
  File "C:\Program Files\Datadog\Datadog Agent\embedded3\lib\site-packages\datadog_checks\base\ddyaml.py", line 47, in safe_yaml_dump_all
    return pyyaml_dump_all(
  File "C:\Program Files\Datadog\Datadog Agent\embedded3\lib\site-packages\yaml\__init__.py", line 241, in dump_all
    dumper.represent(data)
  File "C:\Program Files\Datadog\Datadog Agent\embedded3\lib\site-packages\yaml\representer.py", line 27, in represent
    node = self.represent_data(data)
  File "C:\Program Files\Datadog\Datadog Agent\embedded3\lib\site-packages\yaml\representer.py", line 48, in represent_data
    node = self.yaml_representers[data_types[0]](self, data)
  File "C:\Program Files\Datadog\Datadog Agent\embedded3\lib\site-packages\yaml\representer.py", line 199, in represent_list
    return self.represent_sequence('tag:yaml.org,2002:seq', data)
  File "C:\Program Files\Datadog\Datadog Agent\embedded3\lib\site-packages\yaml\representer.py", line 92, in represent_sequence
    node_item = self.represent_data(item)
  File "C:\Program Files\Datadog\Datadog Agent\embedded3\lib\site-packages\yaml\representer.py", line 58, in represent_data
    node = self.yaml_representers[None](self, data)
  File "C:\Program Files\Datadog\Datadog Agent\embedded3\lib\site-packages\yaml\representer.py", line 231, in represent_undefined
    raise RepresenterError("cannot represent an object", data)
yaml.representer.RepresenterError: ('cannot represent an object', <datadog_checks.base.utils.diagnose.Diagnosis object at 0x0000023555390130>)
```
There is a lot of advice on how to fix it on Google and StackOverflow and I have tried many but failed to find a workable solution so far. It is also architecturally not as ideal as I thought it would be since the Diagnosis object need to "prepare" itself for YAML dumping.

I will talk with @alopezz who is working on Python side if he think it is indeed a superior approach and perhaps he can provide the proper setup for a returned Diagnosis object, than I will make necessary changes elsewhere.

Reply:
There are many ways to go about it, but the simplest thing perhaps would be to have contract of Python's check GetDiagnoses to return JSON with specific keys, and handle the serialization on the Python side entirely. It then would be trivial to plug a custom serializer into JSON.dumps: https://gist.github.com/vickenty/8539570fd291e424c4beffb87cec292c#file-d-py